File: 23.2-completable-future/1-complete/kt/Complete.kt
import java.util.concurrent.CompletableFuture
import kotlin.concurrent.thread

fun delayedValue(value: Int, delay: Long) : CompletableFuture<Int> {
	val res = CompletableFuture<Int>()
	
	thread {
		// after 'delay' milliseconds, complete the future with 'value'
		Thread.sleep(delay)
		res.complete(value)
	}
	
	return res
}

fun main() {
	val future = delayedValue(1234, 3000)

	println(":: future created ::")

	future.thenAccept { value ->
		// will run only after the future is completed normally
		println("++ value produced: ${ value } ++")
	}

	println(":: callback registered ::")

	println(":: waiting... ::")
}

File: 23.2-completable-future/1-complete/js/Complete.js
function delayedValue(value, delay) {
	return new Promise((resolve, reject) => {
		// after 'delay' milliseconds, fulfill the promise with 'value'
		setTimeout(() => { resolve(value) }, delay)
	})
}

const promise = delayedValue(1234, 3000)

console.log(":: promise created ::")

promise.then(value => {
	// will run only after the promise is fulfilled
	console.log(`++ value produced: ${ value } ++`)
})

console.log(":: callback registered ::")

console.log(":: waiting... ::")
